### PR TITLE
refactor: migrate helpers.js/s3-endpoints.js/signing.js to TypeScript

### PR DESCRIPTION
**Pull request by trim21**

### CODE REVIEW

File: src/helpers.ts [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+      throw new errors.InvalidBucketNameError('Invalid Source bucket name: ' + this.Bucket)
+    }
+    if (!isValidObjectName(this.Object)) {
+      throw new errors.InvalidObjectNameError(`Invalid Source object name: ${this.Object}`)
+    }
+    if ((this.MatchRange && this.Start !== -1 && this.End !== -1 && this.Start > this.End) || this.Start < 0) {
+      throw new errors.InvalidObjectNameError('Source start must be non-negative, and start must be at most end.')
+    } else if ((this.MatchRange && !isNumber(this.Start)) || !isNumber(this.End)) {
+      throw new errors.InvalidObjectNameError(
+        'MatchRange is specified. But  Invalid Start and End values are specified. ',
```
Comment (aldy505) [inline src/helpers.ts:100]:
```suggestion
        'MatchRange is specified. But Invalid Start and End values are specified.',
```

Trailing spaces. Just a nit pick.

File: src/internal/helper.ts [hunk 2]
```diff
@@ -0,0 +1,592 @@
...
+
+import type { Binary, Encryption, MetaData, ResponseHeader } from './type.ts'
+import { ENCRYPTION_TYPES } from './type.ts'
+
+/**
+ * All characters in string which are NOT unreserved should be percent encoded.
+ * Unreserved characters are : ALPHA / DIGIT / "-" / "." / "_" / "~"
+ * Reference https://tools.ietf.org/html/rfc3986#section-2.2
+ */
+export function uriEscape(string: string) {
```
Comment (aldy505) [inline src/internal/helper.ts:34]: Should we explicitly declares the return type of the function?

Comment (aldy505) [reply src/internal/helper.ts:34]: Unless @prakashsvmx also agrees that we can let the return types be inferred from usage, then so be it.

Comment (trim21) [reply src/internal/helper.ts:34]: TS can infer return type here. But I'm OK add return type too all functions/methods

File: src/internal/helper.ts [hunk 3]
```diff
@@ -0,0 +1,588 @@
...
+ * @param encConfig
+ * @returns an object with key value pairs that can be used in headers.
+ */
+export function getEncryptionHeaders(encConfig: Encryption): Header {
+  const encType = encConfig.type
+
+  if (!isEmpty(encType)) {
+    if (encType === ENCRYPTION_TYPES.SSEC) {
+      return {
+        [ENCRYPTION_HEADERS.sseGenericHeader]: 'AES256',
```
Comment (trim21) [inline src/internal/helper.ts:507]: https://github.com/minio/minio-js/blob/3ec0a688652fe36678376cc42bff834783019d36/src/helpers.js#L497-L500

I think it's a bug here

**Comment by trim21** [comment]
@prakashsvmx any plan on review this?

File: src/helpers.ts [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+    Bucket,
+    Object,
+    Encryption,
+    UserMetadata,
+    UserTags,
+    LegalHold,
+    RetainUntilDate,
+    Mode,
+  }: {
+    Bucket: string
```
Comment (prakashsvmx) [inline src/helpers.ts:185]: This can be extracted to a type

File: .github/workflows/nodejs-windows.yml [hunk 5]
```diff
@@ -15,7 +15,7 @@ jobs:
     strategy:
       max-parallel: 3
       matrix:
-        node_version: [12.x, 14.x, 16.x, 17.x, 18.x, 20.x]
+        node_version: [10.x, 12.x, 14.x, 16.x, 18.x, 20.x]
```
Comment (prakashsvmx) [inline .github/workflows/nodejs-windows.yml:18]: We recommend LTS versions  
I believe it is ok to add this support for sometime to test in CI for any backward compatibility testing.

File: src/helpers.ts [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+    MatchETag = '',
+    NoMatchETag = '',
+    MatchModifiedSince = null,
+    MatchUnmodifiedSince = null,
+    MatchRange = false,
+    Start = 0,
+    End = 0,
+    Encryption = undefined,
+  }: {
+    Bucket: string
```
Comment (prakashsvmx) [inline src/helpers.ts:64]: This can be extracted into a type definition

File: src/internal/helper.ts (hunk 2 above)
Comment (prakashsvmx) [reply src/internal/helper.ts:35]: We can take it based on community feedback. for now it is fine as is.

File: .github/workflows/nodejs-windows.yml (hunk 5 above)
Comment (trim21) [reply .github/workflows/nodejs-windows.yml:18]: I didn't get it, these are nodejs lts versions

Comment (trim21) [reply .github/workflows/nodejs-windows.yml:18]: do you mean to drop support for end-of-life nodejs LTS version?

https://github.com/nodejs/release#release-schedule

Comment (trim21) [reply .github/workflows/nodejs-windows.yml]: This make things much easier

**Comment by trim21** [comment]
tests on node 12.x are still required.

**Comment by prakashsvmx** [comment]
we can stick to `[12.x, 14.x, 16.x, 17.x, 18.x, 20.x]` for testing.

**Comment by prakashsvmx** [comment]
I am taking a look. :+1:

File: tests/functional/functional-tests.js [hunk 7]
```diff
@@ -2507,7 +2501,7 @@ describe('functional tests', function () {
           if (err) {
             return done(err)
           }
-          if (isArray(tagList)) {
+          if (Array.isArray(tagList)) {
```
Comment (prakashsvmx) [inline tests/functional/functional-tests.js:2504]: a request to update line number  **3775** to
```suggestion
              const stsEndPoint = 'http://localhost:9000'
              const stsEndPoint = 'http://' + clientConfigParams.endPoint + ':' + clientConfigParams.port

```

**Review by prakashsvmx — APPROVED** [review]
Tested. Tests running fine. :+1:

**Review by prakashsvmx — APPROVED** [review]

**Pull request merged**